hathor/feature_activation/feature_service.py

```diff
@@ -12,28 +12,56 @@
 #  See the License for the specific language governing permissions and
 #  limitations under the License.
 
+from typing import TYPE_CHECKING
+
 from hathor.feature_activation.feature import Feature
 from hathor.feature_activation.model.feature_description import FeatureDescription
 from hathor.feature_activation.model.feature_state import FeatureState
 from hathor.feature_activation.settings import Settings as FeatureSettings
-from hathor.transaction import Block
-from hathor.transaction.storage import TransactionStorage
+
+if TYPE_CHECKING:
+    from hathor.transaction import Block
+    from hathor.transaction.storage import TransactionStorage
 
 
 class FeatureService:
     __slots__ = ('_feature_settings', '_tx_storage')
 
-    def __init__(self, *, feature_settings: FeatureSettings, tx_storage: TransactionStorage) -> None:
+    def __init__(self, *, feature_settings: FeatureSettings, tx_storage: 'TransactionStorage') -> None:
         self._feature_settings = feature_settings
         self._tx_storage = tx_storage
 
-    def is_feature_active(self, *, block: Block, feature: Feature) -> bool:
+    def is_feature_active(self, *, block: 'Block', feature: Feature) -> bool:
         """Returns whether a Feature is active at a certain block."""
         state = self.get_state(block=block, feature=feature)
 
         return state == FeatureState.ACTIVE
 
-    def get_state(self, *, block: Block, feature: Feature) -> FeatureState:
+    def check_must_signal(self, block: 'Block') -> bool:
+        """Return whether a block is missing mandatory signaling for any feature currently in the MUST_SIGNAL phase."""
+        bit_counts = block.get_feature_activation_bit_counts()
+        height = block.get_height()
+        offset_to_boundary = height % self._feature_settings.evaluation_interval
+        remaining_blocks = self._feature_settings.evaluation_interval - offset_to_boundary - 1
+        descriptions = self.get_bits_description(block=block)
+
+        must_signal_features = (
+            feature for feature, description in descriptions.items()
+            if description.state is FeatureState.MUST_SIGNAL
+        )
+
+        for feature in must_signal_features:
```

Comment:
I don't get it. Shouldn't we just check if the block has the correct bit signaling for features in the MUST_SIGNAL phase? Why is there a threshold and remaining blocks in the algorithm?

Reply:
This follows the specification in the [RFC](https://github.com/HathorNetwork/rfcs/blob/master/projects/feature-activation/0001-feature-activation-for-blocks.md#invalidation-of-new-blocks-in-must_signal):

> During the `MUST_SIGNAL` state, the threshold must be enforced by invalidating blocks that don't signal support for the feature.
>
> That is, during one evaluation interval, if there are `EVALUATION_INTERVAL - threshold` non-signaling blocks, all new blocks are required to signal support and will be invalidated otherwise.

In other words, what it does is enforcing the same threshold that is used to activate the feature in the first place. For example, if 75% of blocks are required to activate a feature, then we'll tolerate up to 25% of non-signaling blocks during the `MUST_SIGNAL` phase. This is also analogous to Bitcoin's BIP 8.

The reason I see for this is, the `MUST_SIGNAL` phase serves to create a buffer period between a feature getting accepted through voting, and the feature becoming `LOCKED_IN`. Let's say a miner represents 10% of the hash rate, and it votes to _not_ support some feature during its voting phase, that is, it signals `0` for that feature. The feature will have 90% of blocks signaling support (from other miners), so it will eventually become activated. As soon as the voting phase ends, the mechanism shifts to `MUST_SIGNAL`, and all blocks that come from this miner will be rejected. So, we wouldn't be giving him any chance to update his full node and/or enable the signal during the `MUST_SIGNAL` phase.

By allowing for the same threshold used for voting, we give any non-signaling miners time to start signaling before their blocks are rejected. Otherwise, non-signaling miners would have to keep checking if they should enable the signal _before_ it becomes mandatory, which could create a weird game between signaling and non-signaling miners, trying to keep their wish until the last minute before the threshold is reached.

Reply:
Got it. Thanks for the explanation.